### PR TITLE
Allow a list of email domains with Google OAuth 2.0

### PR DESCRIPTION
**Pull request by Misteur-Z**
Rework of #80 
Addressing issue #63 

New features:
* Allows multiple Google OAuth email domains
* Checks for `verified_email` in Google OAuth response
* Log unauthorized attempts as warnings (and not errors)

Sidenotes:
* User `john.doe@domain1.com` and `john.doe@domain2.org` end up with the same jupyterhub user `john.doe` to avoid branking changes

### CODE REVIEW

**Comment by betatim** [comment]
We should continue to support using a string to set a single domain in addition to allowing people to set it to a tuple/list via the configuration. I think it would be nice to maintain backwards compatibility as there is no good reason to break it here.

What motivated your choice to strip the domain names in the case of multiple domains being allowed? Not allowing joe@fake.com to impersonate joe@email.com seems like a desirable feature to me.

File: oauthenticator/google.py [hunk 1]
```diff
@@ -10,8 +10,9 @@
 from tornado import gen
 from tornado.auth import GoogleOAuth2Mixin
 from tornado.web import HTTPError
+from tornado.log import app_log
```
Comment (betatim) [inline oauthenticator/google.py]: this isn't used anywhere

File: oauthenticator/google.py [hunk 2]
```diff
@@ -53,6 +53,11 @@ def _scope_default(self):
 
     @gen.coroutine
     def authenticate(self, handler, data=None):
+        if not self.hosted_domain:
```
Comment (betatim) [inline oauthenticator/google.py]: Not requiring a hosted domain to be configured was a feature I think

**Comment by betatim** [comment]
The docs in the README need updating as well.

File: oauthenticator/google.py (hunk 2 above)
Comment (david-dumas) [reply oauthenticator/google.py]: My logic was you should not allow any domain to login if you setup Google Oauth.

Comment (betatim) [reply oauthenticator/google.py]: I find it convenient that I can use this to require you have a "google" account but I don't actually care you have a particular domain. For example for workshops or other open events where people have all sorts of "google" accounts.

**Comment by david-dumas** [comment]
@betatim I agree maintaining backwards compatibility is better, my python knowledge is limited and I don't know how to solve those two issues cleanly:
* `os.environ.get('HOSTED_DOMAIN', '')` should be converted to a tuple of one string and we have to chose if it overrides `c.GoogleOAuthenticator.hosted_domain` from configuration file
* `hosted_domain` could already be a string or a tuple, the code in `authenticate()` would already work

I will update the README once this is done.

**Comment by david-dumas** [comment]
* Used a List() instead of a Tuple()
* Updated the README
* Backwards compatible

joe@fake.com can impersonate joe@email.com but if we add the user's email domain to it's username [here](https://github.com/jupyterhub/oauthenticator/blob/master/oauthenticator/google.py#L89) to avoid it we create a breaking change...

**Comment by cmseal** [comment]
+1 for this... I had to do something similar, as we've deployed JupyterHub across as group of companies, each with their own domain, but I kept the domain in the username, and just replaced the '@' with another '.' so that joe.bloggs@test.com is different from joe.bloggs@test.co.uk.

**Comment by minrk** [comment]
Thanks for the PR! I added a couple of commits to avoid breaking changes:

1. resolve conflicts
2. only an empty list means no setting, not `['']` (and default HOSTED_DOMAIN='' env results in an empty list now)
3. only strip domain when hosted_domain has only one item, so that collisions cannot occur when allowing multiple domains, like what @cmseal described

**Pull request merged**